fleetspeak/src/server/internal/cache/cache.go

```diff
@@ -0,0 +1,114 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	"github.com/google/fleetspeak/fleetspeak/src/server/db"
+)
+
+// Controls cache timing behavior, variable only to support unit tests.
+var (
+	maxAge         = 30 * time.Second
+	expireInterval = 5 * time.Minute
+)
+
+// Client is a cache of recently connected clients.
+type Client struct {
+	m    map[common.ClientID]*clientEntry
+	l    sync.RWMutex
+	stop chan struct{}
+}
+
+type clientEntry struct {
+	u time.Time
+	d *db.ClientData
+}
+
+// NewClient returns a new cache of client data.
+func NewClient() *Client {
+	ret := &Client{
+		m:    make(map[common.ClientID]*clientEntry),
+		stop: make(chan struct{}),
+	}
+	go ret.expireLoop()
+	return ret
+}
+
+// Get returns the cached client data, if there is sufficiently fresh data in
+// the cache, otherwise nil.
+func (c *Client) Get(id common.ClientID) *db.ClientData {
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	e := c.m[id]
+	if e == nil || db.Now().Sub(e.u) > maxAge {
+		return nil
+	}
+	return e.d
+}
+
+// Update updates or sets the cached data for a particular client. If data is
+// nil, it clears the data for the client.
+func (c *Client) Update(id common.ClientID, data *db.ClientData) {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	if data == nil {
+		delete(c.m, id)
+	} else {
+		c.m[id] = &clientEntry{
+			u: db.Now(),
+			d: data,
+		}
+	}
+}
+
+// Clear empties the cache, removing all entries.
+func (c *Client) Clear() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	c.m = make(map[common.ClientID]*clientEntry)
+}
+
+// Stop releases the resources required for background cache maintenence. The
+// Cache should not be used once Stop has been called.
+func (c *Client) Stop() {
+	close(c.stop)
+}
+
+// Size returns the current size taken up by the cache, this is a count of
+// client records, some of which may be too old to be visible.
+func (c *Client) Size() int {
+	c.l.RLock()
+	defer c.l.RUnlock()
+	return len(c.m)
+}
+
+func (c *Client) expireLoop() {
+	t := time.NewTicker(expireInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			c.expire()
+		case <-c.stop:
+			return
+		}
+	}
+}
+
+// expire prunes the cache to clean out clients that are no longer up to date.
+func (c *Client) expire() {
+	c.l.Lock()
+	defer c.l.Unlock()
+
+	for k, e := range c.m {
+		if db.Now().Sub(e.u) > maxAge {
+			delete(c.m, k)
```

Comment:
Are you sure modifying maps while iterating over them is allowed? In Python it raises.

Reply:
It is fine. https://stackoverflow.com/questions/23229975/is-it-safe-to-remove-selected-keys-from-golang-map-within-a-range-loop